scipy/_lib/tests/test__util.py

```diff
@@ -349,3 +352,46 @@ def mock_str():
     ]
 
     assert res == ref
+
+
+class TestLazywhere:
+    # Mixing floats and integers does not work in np.array_api, so don't try
+    n_arrays = strategies.integers(min_value=1, max_value=3)
+    rng_seed = strategies.integers(min_value=1000000000, max_value=9999999999)
+    dtype = strategies.sampled_from((np.float32, np.float64))
+    p = strategies.floats(min_value=0, max_value=1)
+    data = strategies.data()
+
+    @pytest.mark.filterwarnings('ignore::RuntimeWarning')  # overflows, etc.
+    @array_api_compatible
```

Comment:
```suggestion
    @skip_if_array_api_gpu
    @array_api_compatible
```
Consider adding per https://github.com/scipy/scipy/pull/18930#discussion_r1300703516 (with required import at the top). 

Alternatively, since array API behavior is opt-in and CI doesn't use GPUs right now, leave as-is. A failure in the future will prompt the use of the a GPU-compatible `assert_equal` (which will presumably have been written by the time anyone sees a failure).

Reply:
@lucascolley IIUC this test should now pass with CuPy arrays, so we can resolve this?

Reply:
Looks fine with CuPy. `python dev.py test -t scipy/_lib/tests/test__util.py -b all` gives me this error with `torch`:
<details>

```
scipy/_lib/tests/test__util.py:364: in test_basic
    @array_api_compatible
        f          = <function given.<locals>.run_test_as_given.<locals>.wrapped_test at 0x7f22e371e980>
        self       = <scipy._lib.tests.test__util.TestLazywhere object at 0x7f22e3d9bb50>
        xp         = <module 'torch' from '/home/lucas/mambaforge/envs/scipy-dev/lib/python3.11/site-packages/torch/__init__.py'>
scipy/_lib/tests/test__util.py:399: in test_basic
    xp_assert_equal(res1, ref1)
        arrays     = [tensor(0.), tensor(nan)]
        cond       = tensor(True)
        cond_shape = ()
        data       = data(...)
        dtype      = <class 'numpy.float32'>
        f          = <function TestLazywhere.test_basic.<locals>.f at 0x7f2342606e80>
        f2         = <function TestLazywhere.test_basic.<locals>.f2 at 0x7f234263a3e0>
        fillvalue  = tensor(0.)
        input_shapes = ((), (), ())
        mbs        = mutually_broadcastable_shapes(num_shapes=3, min_side=0)
        n_arrays   = 2
        p          = 0.0
        ref1       = tensor(nan)
        ref2       = tensor(nan)
        res1       = tensor(nan)
        res2       = tensor(nan)
        result_shape = ()
        rng        = Generator(PCG64) at 0x7F234263E5E0
        rng_seed   = 1000000000
        self       = <scipy._lib.tests.test__util.TestLazywhere object at 0x7f22e3d9bb50>
        shapes     = [(), ()]
        xp         = <module 'torch' from '/home/lucas/mambaforge/envs/scipy-dev/lib/python3.11/site-packages/torch/__init__.py'>
scipy/_lib/_array_api.py:188: in assert_equal
    return xp.testing.assert_close(actual, desired, rtol=0, atol=0,
E   AssertionError
E   Falsifying example: test_basic(
E       self=<scipy._lib.tests.test__util.TestLazywhere object at 0x7f22e3d9bb50>,
E       xp=<module 'torch' from '/home/lucas/mambaforge/envs/scipy-dev/lib/python3.11/site-packages/torch/__init__.py'>,
E       n_arrays=2,
E       rng_seed=1000000000,
E       dtype=numpy.float32,
E       p=0.0,
E       data=data(...),
E   )
E   Draw 1: BroadcastableShapes(input_shapes=((), (), ()), result_shape=())
E   Draw 2: array(0., dtype=float32)
E   Draw 3: array(0., dtype=float32)
E   Draw 4: array(nan, dtype=float32)
E   
E   You can reproduce this example by temporarily adding @reproduce_failure('6.82.7', b'AXicY2RgZCANKDAy/P8BYQIAEN8CGw==') as a decorator on your test case
        actual     = tensor(nan)
        desired    = tensor(nan)
        err_msg    = ''
        xp         = <module 'scipy._lib.array_api_compat.array_api_compat.torch' from '/home/lucas/dev/myscipy/build-install/lib/python3.11/site-packages/scipy/_lib/array_api_compat/array_api_compat/torch/__init__.py'>
========================================================================================= short test summary info ==========================================================================================
FAILED scipy/_lib/tests/test__util.py::TestLazywhere::test_basic[torch] - AssertionError
====================================================================================== 1 failed, 24 passed in 26.03s =======================================================================================
```

</details>

Reply:
Thanks. Can you try again after implementing https://github.com/scipy/scipy/pull/19186/files#r1316228638.

Reply:
`python dev.py test -t scipy/_lib/tests/test__util.py -b all` passes all tests after that patch yup.

However, with `SCIPY_DEVICE=cuda`, I get this error:

<details>

```
____________________________________________________________________________________ TestLazywhere.test_basic[torch] ______________________________________________________________________________________
scipy/_lib/tests/test__util.py:364: in test_basic
    @array_api_compatible
        f          = <function given.<locals>.run_test_as_given.<locals>.wrapped_test at 0x7f9b0639fc40>
        self       = <scipy._lib.tests.test__util.TestLazywhere object at 0x7f9b066ba390>
        xp         = <module 'torch' from '/home/lucas/mambaforge/envs/scipy-dev/lib/python3.11/site-packages/torch/__init__.py'>
scipy/_lib/tests/test__util.py:384: in test_basic
    res1 = _lazywhere(cond, arrays, f, fillvalue)
        arrays     = [tensor([])]
        cond       = tensor(True)
        cond_shape = ()
        data       = data(...)
        dtype      = <class 'numpy.float32'>
        f          = <function TestLazywhere.test_basic.<locals>.f at 0x7f9be61a6de0>
        f2         = <function TestLazywhere.test_basic.<locals>.f2 at 0x7f9b652c2020>
        fillvalue  = tensor(0.)
        input_shapes = ((), (0,))
        mbs        = mutually_broadcastable_shapes(num_shapes=2, min_side=0)
        n_arrays   = 1
        p          = 0.0
        result_shape = (0,)
        rng        = Generator(PCG64) at 0x7F9B652E5FC0
        rng_seed   = 1000000000
        self       = <scipy._lib.tests.test__util.TestLazywhere object at 0x7f9b066ba390>
        shapes     = [(0,)]
        xp         = <module 'torch' from '/home/lucas/mambaforge/envs/scipy-dev/lib/python3.11/site-packages/torch/__init__.py'>
scipy/_lib/_util.py:119: in _lazywhere
    out[cond] = temp1
E   RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
```

</details>

Since `cond` is created from a `numpy` array (`cond = xp.asarray(rng.random(size=cond_shape) > p)`), it is created on the cpu.

Reply:
> ... it is created on the cpu.

Do I understand that:
- `cond = xp.asarray(rng.random(size=cond_shape) > p)` generates an array on the CPU
- `arrays = [xp.asarray(data.draw(npst.arrays(dtype=dtype, shape=shape)))
                  for shape in shapes]` also generates arrays on the CPU
- `xp = array_namespace(cond, *arrays)` returns a Pytorch namespace
- `out = xp.full(cond.shape, fill_value=fillvalue, dtype=dtype)` generates an array on a different device.

This sounds like it is going to be a common gotcha. I would not expect the namespace returned by `array_namespace` to return functions that will produce arrays that are incompatible with the arrays that were provided as arguments to `array_namespace`.

Reply:
I think that may be the case, but I don't know how `asarray` is supposed to interact with the default device and whether the fix linked above will apply to `numpy` array inputs too.

The ideal scenario is that `xp.asarray` transfers to GPU à la CuPy when we have `SCIPY_DEVICE=cuda`.

I'll let you know once I hear back from PyTorch and am able to test with the nightly.

Reply:
@mdhaber crisis averted: the https://github.com/pytorch/pytorch/issues/106773 fix applies to `np` inputs too, so with a recent `pytorch` nightly `cond = xp.asarray(rng.random(size=cond_shape) > p)` will generate a GPU array when we have `SCIPY_DEVICE=cuda`.

Reply:
Cool. Thanks.

Reply:
> So everything in this PR is looking good

Thanks. In that case, please consider hitting the approve button. (As a maintainer, it's helpful to see that someone else has also taken a look.)